File: Project Enhancements/Software Design & Engineering/Brea_SoftDesignEngineeringEnhancement/Source/src/Task.cpp
#include "Task.h"
#include <stdexcept>

Task::Task(const std::string& id, const std::string& name, const std::string& description)
    : id(id), name(name), description(description) {

    //validate ID
    if (id.empty() || id.length() > 10) {
        throw std::invalid_argument("Invalid ID");
    }

    //validate other member variables
    validateName(name);
    validateDescription(description);
}


//return task id
std::string Task::getID() const {
    return id;
}

//return task name
std::string Task::getName() const {
    return name;
}

//return task description
std::string Task::getDescription() const {
    return description;
}

//set new task name
void Task::setName(const std::string& name) {
    validateName(name);
    this->name = name;
}

//set new description
void Task::setDescription(const std::string& description) {
    validateDescription(description);
    this->description = description;
}

//make sure name meets required input guidelines
void Task::validateName(const std::string& name) {
    if (name.empty() || name.length() > 20) {
        throw std::invalid_argument("Invalid Name");
    }
}

//make sure description meets required input guidelines
void Task::validateDescription(const std::string& description) {
    if (description.empty() || description.length() > 50) {
        throw std::invalid_argument("Invalid Description");
    }
}

File: Project Enhancements/Databases/src/src/Task.java
package src;

public class Task {

	private final String ID;
	private String name;
	private String desc;
	
	/**
	 * @param iD
	 * @param name
	 * @param description
	 */
	public Task(String ID, String name, String desc) {
		//checks if ID is null, more than 10 characters, or empty
		if (ID == null || ID.length() > 10 || ID.isEmpty()) {
			throw new IllegalArgumentException("Invalid ID");
		}
		
		//checks if Name is null, more than 20 characters, or empty
		if (name == null || name.length() > 20 || name.isEmpty()) {
			throw new IllegalArgumentException("Invalid Name");
		}
		
		//checks if Description is null, more than 50 characters, or empty
		if (desc == null || desc.length() > 50 || desc.isEmpty()) {
			throw new IllegalArgumentException("Invalid Description");
		}
		
		//set all if nothing thrown
		this.ID = ID;
		this.name = name;
		this.desc = desc;
	}
	
	/**
	 * @return the name
	 */
	public String getName() {
		return name;
	}
	
	/**
	 * @param name the name to set
	 */
	public void setName(String name) {
		if (name == null || name.length() > 20 || name.isEmpty()) {
			throw new IllegalArgumentException("Invalid Name");
		}
		
		this.name = name;
	}
	
	/**
	 * @return the description
	 */
	public String getDescription() {
		return desc;
	}
	
	/**
	 * @param description the description to set
	 */
	public void setDescription(String desc) {
		if (desc == null || desc.length() > 50 || desc.isEmpty()) {
			throw new IllegalArgumentException("Invalid Description");
		}
		
		this.desc = desc;
	}
	
	/**
	 * @return the iD
	 */
	public String getID() {
		return ID;
	}
	
	
}
